src/PhpDocParser/PhpStan.php

```diff
@@ -0,0 +1,159 @@
+                $type = $node->type->type;
+
+                /** @var ReflectionClass $class */
+                $class = $parameter->getDeclaringClass();
+                /** @var string $fileName */
+                $fileName = $class->getFileName();
+                /** @var string $phpFileContent */
+                $phpFileContent = file_get_contents($fileName);
+                /** @var Stmt[] $parsedFile */
+                $parsedFile = $this->phpParser->parse($phpFileContent);
+
+                $namespace = $this->getNamespaceStmt($parsedFile);
+                $uses = $this->getUseStmts($namespace);
+                $namespaces = $this->getUsesNamespaces($uses);
+
+                return $this->getFullClassName($type->name, $namespaces, $class);
+            }
+        }
+
+        throw new BuildingError();
+    }
+
+    private function isScalar(string $value): bool
+    {
+        $scalars = [
+            'string',
+            'bool',
+            'int',
+            'float',
+            'double',
+            'mixed',
+        ];
+
+        return in_array($value, $scalars, true);
+    }
+
+    /**
+     * @param Stmt[] $nodes
+     */
+    private function getNamespaceStmt(array $nodes): Stmt\Namespace_
+    {
+        foreach ($nodes as $node) {
+            if ($node instanceof Stmt\Namespace_) {
+                return $node;
+            }
+        }
+
+        return new Stmt\Namespace_();
+    }
+
+    /** @return Stmt\Use_[] */
+    private function getUseStmts(Stmt\Namespace_ $node): array
+    {
+        return array_filter($node->stmts, function (Stmt $node): bool {
+            return $node instanceof Stmt\Use_;
+        });
+    }
+
+    /**
+     * @param Stmt\Use_[] $uses
+     * @return string[]
+     */
+    private function getUsesNamespaces(array $uses): array
+    {
+        return array_map(function (Stmt\Use_ $use): string {
+            return $use->uses[0]->name->toString();
+        }, $uses);
+    }
+
+    /** @param string[] $namespaces */
+    private function getFullClassName(string $name, array $namespaces, ReflectionClass $class): string
+    {
+        if ('\\' === $name[0] || explode('\\', $name)[0] !== $name) {
+            return $name;
+        }
+
+        if (0 === count($namespaces)) {
+            return '\\' . $class->getNamespaceName() . '\\' . $name;
+        }
+
+        return '\\' . $this->getNamespaceForClass($name, $namespaces);
+    }
+
+    /**
+     * @param string[] $namespaces
+     * @throws BuildingError
+     */
+    private function getNamespaceForClass(string $className, array $namespaces): string
+    {
+        foreach ($namespaces as $namespace) {
+            if ($this->endsWith($namespace, $className)) {
+                return $namespace;
+            }
+        }
+
+        throw new BuildingError('Can not resolve namespace for class ' . $className);
+    }
+
+    private function endsWith(string $haystack, string $needle): bool
```

Comment:
Maybe move it to function or trait

Reply:
Not now, because I will refactor `Builder\Reflection` to use `PhpDocParser\PhpStan`